src/nubank/gen_cert.rs

```diff
@@ -0,0 +1,216 @@
+use openssl::pkcs12::Pkcs12;
+use openssl::pkey::{PKey, Private};
+use openssl::rsa::Rsa;
+use openssl::x509::X509;
+use random_string::generate;
+use serde::Deserialize;
+use serde_json::{self};
+use std::collections::HashMap;
+use std::fs;
+use std::str::{self, Utf8Error};
+#[path = "./discovery.rs"]
+mod discovery;
+
+#[derive(Debug)]
+pub enum GenCertError {
+    DiscoveryError(discovery::DiscoveryError),
+    FailedToGeneratePrivateKey(openssl::error::ErrorStack),
+    FailedToGeneratePublicKey(openssl::error::ErrorStack),
+    FailedToConvertPublicKeyToStr(Utf8Error),
+    MainRequestFailed(reqwest::Error),
+    HeaderKeyNotFound(reqwest::header::HeaderMap),
+    FailedToConvertHeaderValueToStr(reqwest::header::ToStrError),
+    FailedToReadHeaderValue(String),
+    HeaderValueKeyNotFound(reqwest::header::HeaderValue),
+    ExchangeCertRequestFailed(reqwest::Error),
+    ExchangeCertRequestDecodingFailed(reqwest::Error),
+    ExchangeCertJsonConversionFailed(serde_json::Error),
+    FailedToGenerateCert(openssl::error::ErrorStack),
+    FailedToWriteCert(std::io::Error),
+}
+
+#[derive(Debug, Clone)]
+pub struct CodeRequestOutput {
+    url: String,
+    pub sent_to: String,
+    encrypted_code: String,
+    payload: HashMap<String, String>,
```

Comment:
Why is it a HashMap instead of a struct? structs are safer and more efficient

Reply:
For me, the rule of thumb is to use structs when the field names are known and HashMap otherwise

Reply:
I thought that `reqwest` library would accept only HashMaps as input for the payload. Reading the documentation you can see it accepts everything that implements the serialize trait.

Reply:
Great, so implement it